### PR TITLE
Fix infinite loop with using KeyboardInset

### LINKED ISSUE #31: Infinite recursion in the KeyboardInsetHelper.
@stephanecopin we have a critical issue in the KeyboardInsetHelper: infinite recursion call:
```
	@available(*, deprecated, renamed: "updateForInset(_:base:)")
-->	@objc open func updateForInset(_ inset: CGFloat) {
		self.updateForInset(inset, base: inset)
	}
///...///
	open func updateForInset(_ inset: CGFloat, base baseInset: CGFloat) {
		// Just to avoid the deprecation warning that would otherwise display... and it's required for backward compatibility
-->		self.perform(#selector(NoDeprecationWarningsHelper.updateForInset(_:)), with: inset as NSNumber)
		scrollView?.contentInset.bottom = inset
		scrollView?.scrollIndicatorInsets.bottom = inset
		constraint?.constant = inset
		referenceView?.layoutIfNeeded()
	}
```
The current implementation can't be used because of that.

### PR DESCRIPTION
**Pull request by stephanecopin**
Fix #31

### CODE REVIEW

File: FueledUtils/KeyboardInsetHelper.swift [hunk 1]
```diff
@@ -129,7 +133,9 @@ open class KeyboardInsetHelper: NSObject {
 	///
 	open func updateForInset(_ inset: CGFloat, base baseInset: CGFloat) {
 		// Just to avoid the deprecation warning that would otherwise display... and it's required for backward compatibility
+		self.isCallingDeprecatedMethod = true
 		self.perform(#selector(NoDeprecationWarningsHelper.updateForInset(_:)), with: inset as NSNumber)
+		self.isCallingDeprecatedMethod = false
```
Comment (heymansmile) [inline FueledUtils/KeyboardInsetHelper.swift:138]: @stephanecopin a flag just to avoid a deprecation warning?! Why do you need this call at all? Why not just:
```
@available(*, deprecated, renamed: "updateForInset(_:base:)")
open func updateForInset(_ inset: CGFloat) {
	self.updateForInset(inset, base: inset)
}

open func updateForInset(_ inset: CGFloat, base baseInset: CGFloat) {
	scrollView?.contentInset.bottom = inset
	scrollView?.scrollIndicatorInsets.bottom = inset
	constraint?.constant = inset
	referenceView?.layoutIfNeeded()
}
```

**Review by heymansmile — CHANGES_REQUESTED** [review]

File: FueledUtils/KeyboardInsetHelper.swift (hunk 1 above)
Comment (stephanecopin) [reply FueledUtils/KeyboardInsetHelper.swift:138]: @heymansmile I'll explain my whole way of thinking here (also so everybody understands)
If this project was private, I'd agree with you - I'd make sense to do it the way you've mentioned, but yet what you've mentioned could create sneaky issues in the projects that use FueledUtils (and I _know_ it would, in fact, which is the reason why I've done it the way I did).

In your snippet, the big difference between what I've done and what you have is that, you didn't take into account that the new `KeyboardInsetHelper` now **only** calls `updateForInset(_:, base: CGFloat)`. This means that old code, that relied on subclassing `KeyboardInsetHelper` and override the method `updateForInset(_:)` would **stop working** - because with your snippet, this method wouldn't get called anymore after updating. This could create issues that potentially aren't easy to see (and I know at least 2 projects I've worked on that are relying on this).

My code takes this into account: if you were overriding `updateForInset(_:)`, updating will maintain backward compatibility. It's also important that deprecation warnings are just that, i.e. that they keep the code working as it was, yet indicating you you should update something. Your snippet gives the warning but yet breaks the previous behavior.

I'll also add that maintaining backward compatibility is important regarding to semantic versioning:
- If breaking backward compatibility, the library should increase its major version
- If not, but adding features, it should increase its minor version
(and if just bug fixes without interfaces change, its patch version)

I agree it's my mistake I didn't test the code and account for the infinite loop right away, but to be fair you and other people reviewed the relevant PR and approved it. Errare humanum est 🤷‍♂️ 

TL;DR maintaining backward compatibility is important and your snippet doesn't

Comment (heymansmile) [reply FueledUtils/KeyboardInsetHelper.swift:138]: @stephanecopin 
- Actually all my attention was pointing to the infinite loop bug and your proposed fix of that with the flag. I missed the fact that now `KeyboardInsetHelper` is open class instead of final and `updateForInset` is open instead of public how they were created originally. 
**Anyway** your solution breaks backward compatibility in this simple case:
```
class CustomKeyboardInsetHelperEXISTING: KeyboardInsetHelper {
	override func updateForInset(_ inset: CGFloat) {
		//...the super class functionality will be called anyway which is wrong...//
	}
}
```

- I don't see a problem in removing the old method with previous signature. We've made it open and started our own way with the Utilities and it's fine to move forward with it instead of adding bugs and fragile workarounds.

- Regarding to the fact that the PR was approved by me and others: it was a 74-modified-files-long PR with a name "**Document FueledUtils**" where me and you had some long conversations btw. It was easy to skip such changes.

Comment (stephanecopin) [reply FueledUtils/KeyboardInsetHelper.swift:138]: - It's been `open` since [October 25, *2017*](https://github.com/Fueled/ios-utilities/commit/60ce9956c1a766190b2e5491164dfe93914f7c21). That's a _long time_ for people to use and override this class and method in a lot of projects.
- It indeed does, however I don't believe anybody ever did that. My fix _at least_ preserve functionality in the way I know it was used, while allowing more usage that people would need (as I know I did in the past). If people override this method, I will assume that none of the outlets are bound, making the original implementation irrelevant anyway.
- Breaking backward compatibility is not okay, especially since this is open source.
- In the description of the PR, I did mention it was also deprecating methods while trying to maintain backward compatibility. I could have done a better job indeed and detailed every single change, but I would have thought people would have been more thorough. My bad here.

All in all, please approve this PR so we can get the fix merged.

Comment (heymansmile) [reply FueledUtils/KeyboardInsetHelper.swift:138]: _Breaking backward compatibility is not okay, especially since this is open source._  - again... you started with such solution which breaks backward compatibility, I just asked why this exact solution (because using flag & performig selector wasn't clear enough for me). After that you told in everyone's hearing that I'm the bad guy who displays bad practices by proposing breaking backward compatibility 😖 

Well.. @stephanecopin I don't agree with your solution and your approaches in discussing things but I will approve to get the bug fixed.

**Review by heymansmile — APPROVED** [review]
